_source/_posts/2022-01-18-deploy-dotnet-container-aws-fargate.md

```diff
@@ -0,0 +1,640 @@
+
+To talk about containers, we have to talk about Virtual Machines. A virtual machine (VM) is a virtual representation of a computer all the way from its boot process to loading an entire operating system. This provides a lot of flexibility because you gain fine-grained control over an entire virtual computer, but it brings all the pitfalls of a whole computer as well. The operating system has to be maintained, patched, and updated; disks have to be managed, and all manner of operational overhead goes into their care and feeding. Notably, VMs are also huge because an entire OS has to go on them. So when you start a VM, you have to go through the entire boot sequence of a traditional computer.
+
+These are the three problem areas that containers hope to solve; Ease of management, size, and speed to start. There are others, but for simplicity, we will focus on those three as I feel (IMHO) they are the most relevant to business processes. Containers differ from a traditional virtual machine in one significant way. They do not have an actual Operating System. This may come as a surprise to you, even if you might have dabbled with containers before. Containers instead have an abstraction of an operating system that provides hooks into the host operating system using standards and conventions supplied by the container system.
+
+Docker is the de facto container standard for all practical purposes, but other container technologies do exist. The notion of a standardized container comes from the shipping industry. In shipping, there was no standardization, which made transport difficult because you had to figure out how to hold each item on a boat, train, or truck. Containers provided a standard format for moving goods "you can put whatever you want inside the container so long as it conforms to a standard set of dimensions and can be opened and bolted down in a uniform way." This simplified the logistical process of moving real-world items. In this same way, standardizing the way a software container interacts with its host OS to delegate the responsibility for executing tasks simplifies management of the most crucial part, the application.
+
+## Hosting Containers with AWS ECS Fargate
+
+Now that we have a better idea of what makes a container different from a virtual machine, how does this solve our three problems?
+
+First, it makes management easier by letting us write build scripts to create a stable, repeatable representation of the host needed to do the body of work needed by the application. Second, since an OS ultimately hosts the container, that OS can enforce company policy and security policies, so at worst, any container can only do at most that much. For example, if the host OS only allows traffic inbound from a specific subnet and port, the container cannot override that as it is ultimately bound by the host OS's networking rules.
```

Comment:
```suggestion
First, containers make management easier by letting us write build scripts to create a stable, repeatable representation of the host. This is needed to do the body of work required by the application. Second, since an OS ultimately hosts the container, that OS can enforce company and security policies.  In the worst case, this would be the most a given container could do. For example, if the host OS only allows traffic inbound from a specific subnet and port, the container cannot override this restriction. The container is ultimately bound by the host OS's networking rules.
```
simplifying complex multi-clause sentences for reabability